docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+```mermaid
+classDiagram
+
+Keyring <|-- Wallet
+
+SecureStorage <|-- Keyring
+SecureItem <|-- SecureStorage
+CryptoProvider <|-- SecureItem
+
+Hasher <|-- CryptoProvider
+Verifier <|-- CryptoProvider
+Signer <|-- CryptoProvider
+Cipher <|-- CryptoProvider
+Generator <|-- CryptoProvider
+```
+
+### Crypto Provider
+
+*Crypto Providers* serve as a middleware responsible for managing the interaction with various instantiated cryptographic packages. It acts as a centralized controller, encapsulating the API of the crypto modules in a single location.
+Through each Crypto provider, users can access functionality such as signing, verification, encryption, and hashing.
+
+By abstracting the underlying cryptographic functionality, *Crypto providers* enable a modular and extensible architecture. It allows users to easily switch between different cryptographic implementations without impacting the rest of the system.
+
+A `CryptoProvider` can be built from different sources, such as a `SecureItem`, a randomness source, a seed, a mnemonic, or a string.
+Since the underlying representation of a `CryptoProvider` is a protobuf message, it can be easily serialized and stored in a safe manner.
+The storing and retrieval is handled by a `SecureStorage` implementation which will be presented in the next section.
+
+```go
+type ProviderMetadata interface {
+  key string
+  value string
+}
+
+type ICryptoProviderMetadata interface {
+  GetTypeUUID() TypeUUID
+  GetName() string
+  GetMetadata() []ProviderMetadata
+}
+
+type ICryptoProviderBuilder interface {
+  ICryptoProviderMetadata
+  
+  FromSecureItem( item SecureItem ) (ICryptoProvider, error)  
+
+  FromRandomness( source IRandomnessSource ) (ICryptoProvider, error)
+  FromSeed( seed []byte ) (ICryptoProvider, error)
+  FromMnemonic( mnemonic string ) (ICryptoProvider error)
+  FromString( url string ) (ICryptoProvider error)
+}
+
+type ICryptoProvider interface {
+  Proto.Message
+  ICryptoProviderMetadata
+  
+  GetKeys() (PubKey, PrivKey, error)
+  GetSigner() (ISigner, error)
+  GetVerifier() (IVerifier, error)
+  GetCipher() (ICipher, error)
+  GetHasher() (IHasher, error)
+}
+```
+
+#### Signing
+
+Interface responsible for signing a message and returning the generated signature.
+
+```go
+type ISigner interface {
+  Sign(Blob) (Signature, error)
+}
+```
+
+#### Verifier
+
+Verifies if given a message belongs to a public key by validating against its respective signature.
+
+```go
+type IVerifier interface {
+  Verify(Blob, Signature) (bool, error)
+}
+```
+
+#### Cipher
+
+A cipher is an api for encryption and decryption of data. Given a message it should operate through a secret.
+
+```go
+type ICipher interface {
+  Encrypt(message Blob) (encryptedMessage Blob, error)
+  Decrypt(encryptedMessage Blob) (message Blob, error)
+}
+```
+
+#### Hasher
+
+This package contains the different hashing algorithms and conventions agreed on this matter.
+
+```go
+type IHasher interface {
```

Comment:
Why does hashing need to be abstracted into the CryptoProvider? It involves no private keys, and is basically never delegated to hardware modules.

Reply:
This is not the case in hardware wallets, that receive a plain tx (blob) to sign and not the digest itself. Sending complete blobs (non hashed) to providers allows them to add access control rules, reject/allow based on content, etc